python-api/gstswitch/server.py

```diff
@@ -22,6 +24,107 @@
 TOOLS_DIR = '/'.join(os.getcwd().split('/')[:-1]) + '/tools/'
 
 
+class ProcessMonitor(object):
+    """Runs a Command in a Background-Thread and monitors it's output
+
+    Can block until the Command prints a certain string and log the full
+    output into a file
+    """
+
+    def __init__(self, cmd, logtarget=sys.stderr):
+        self.log = logging.getLogger('server-output-monitor')
+
+        # After calling start(), _proc contains a subprocess.Popen instance
+        self._proc = None
+
+        # Command to run (Array passed to subprocess.Popen)
+        self._cmd = cmd
+
+        # Logfile to write to
+        self._logtarget = logtarget
+
+        # Internal Buffer to search for matched when wait_for_output is called
+        self._buffer = ""
+
+    def start(self):
+        """Starts up the process specified in the constructor in a thread.
+        Blocks until the process is started and has a pid assigned, but not
+        longer then 0.5s.
+        """
+
+        assert self._proc is None
+        self.log.debug("starting subprocess")
+
+        self._proc = subprocess.Popen(
+            self._cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            bufsize=0,
+            shell=False)
+
+        assert self._proc is not None
+        self.log.debug("subprocess successfully")
+
+    def terminate(self):
+        """Kills the process and waits for the thread to exit"""
+        self.log.debug("killing the subprocess")
+        if self._proc:
+            self._proc.terminate()
+
+    def poll(self):
+        """Calls poll() on the running process and returns its result or None
+        if no process is running"""
+        if self._proc:
+            return self._proc.poll()
+
+        return None
+
+    @property
+    def pid(self):
+        """Returns the pid of the running process or None
+        if no process is running"""
+        if self._proc:
+            return self._proc.pid
+
+        return None
+
+    def wait_for_output(self, match, timeout=5, count=1):
```

Comment:
I think it would be better if there was a `check_for_output` method which was non-blocking which was then called by `wait_for_output`.

```python
def check_for_output(self, match):
   xxxx

def wait_for_output(self, match, timeout=5):
  start = time.time()
  while time.time() - start < timeout:
    if self.check_for_output(match):
      return True
    time.sleep(0.1)
  raise TimeoutError("Didn't find %s" % match)
```